lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp

```diff
@@ -437,29 +858,69 @@ void HdVP2Mesh::Sync(
 #endif
     }
 
-    if (HdChangeTracker::IsPrimvarDirty(*dirtyBits, id, HdTokens->normals)
+    if (HdChangeTracker::IsPrimvarDirty(*dirtyBits, id, HdTokens->points)
+        || HdChangeTracker::IsPrimvarDirty(*dirtyBits, id, HdTokens->normals)
         || HdChangeTracker::IsPrimvarDirty(*dirtyBits, id, HdTokens->primvar)) {
-        const HdVP2Material* material = static_cast<const HdVP2Material*>(
-            renderIndex.GetSprim(HdPrimTypeTokens->material, GetMaterialId()));
 
-        const TfTokenVector& requiredPrimvars = material && material->GetSurfaceShader()
-            ? material->GetRequiredPrimvars()
-            : sFallbackShaderPrimvars;
+        auto addRequiredPrimvars = [&](const SdfPath& materialId) {
+            const HdVP2Material* material = static_cast<const HdVP2Material*>(
+                renderIndex.GetSprim(HdPrimTypeTokens->material, materialId));
+            const TfTokenVector& requiredPrimvars = material && material->GetSurfaceShader()
+                ? material->GetRequiredPrimvars()
+                : sFallbackShaderPrimvars;
+
+            for (const auto& requiredPrimvar : requiredPrimvars) {
+                if (!_PrimvarIsRequired(requiredPrimvar)) {
+                    _meshSharedData->_allRequiredPrimvars.push_back(requiredPrimvar);
+                }
+            }
+        };
+
+        // there is a chance that the geom subsets cover all the faces of the
+        // mesh and that the overall material id is unused. I don't figure that
+        // out until much later, so for now just accept that we might pull unnecessary
+        // primvars required by the overall material but not by any of the geom subset
+        // materials.
+        addRequiredPrimvars(GetMaterialId());
+
+        // If we are drawing GeomStyleHull there there could be geom subset material
+        // bindings. Add their additional required primvars to the overall list
+        _MeshReprConfig::DescArray reprDescs = _GetReprDesc(reprToken);
+
+        // For each relevant draw item, update dirty buffer sources.
+        bool hasHullGeomStyle = false;
+        for (size_t descIdx = 0; descIdx < reprDescs.size(); ++descIdx) {
+            const HdMeshReprDesc& desc = reprDescs[descIdx];
+            if (desc.geomStyle == HdMeshGeomStyleHull) {
+                hasHullGeomStyle = true;
+                break;
+            }
+        }
```

Comment:
Can Hydra invalidate primvars when we switch from wireframe mode to shaded mode in the viewport? If not, we will have to always add primvar requirements for each bound material regardless of the state of `hasHullGeomStyle`.

Reply:
I tried it out and you are correct. If I load the scene in wireframe mode & switch over to shaded then VP2 complains about missing buffers. I guess the primvars are not dirty and so we don't re-run this code. We'd have to always compare the old geometry requirements & new geometry requirements to see if they change and then force extraction.

I will switch it to always include all the requirements for now and we can choose to optimize wireframe by supporting fewer requirements later.

Reply:
We add the base material id primvars already anyway, so this is not very different.